### PR TITLE
Removes annotate transformations for loops and branches

### PR DESCRIPTION
**Pull request by phschaad**
- Remove the need to call `apply_transformations_repeated` on the SDFG when propagating states/memlets to reduce time complexity of the algorithm
    - Removes Loop Annotation transformation
    - Removes Branch Annotation transformation
- Ensure fully merged branches are detected within loops (Add test)
- Reduce the amount of dominance and loop calculations on the graph
- Ensure loop guards inside branches get the correct number of executions (Add test)

### CODE REVIEW

**Review by tbennun — CHANGES_REQUESTED** [review]
Please refactor and correct DF as discussed.

File: dace/sdfg/propagation.py [hunk 1]
```diff
@@ -553,16 +554,171 @@ def propagate(self, array, expressions, node_range):
...
+    nontree_edges = [e for e in labelle_edges if e[2] == 'nontree']
+    back_edges = []
+
+    # Given all non-tree edges, which are either forward, cross-edges, or back-
+    # edges, identify all back-edges.
+    for nontree_edge in nontree_edges:
+        edge = sdfg.edges_between(nontree_edge[0], nontree_edge[1])[0]
+        if (nx.has_path(sdfg.nx, edge.dst, edge.src)):
+            # A non-tree edge is a back edge if there is a path from the edge's
+            # destination to it's source. We temporarily remove this.
```
Comment (tbennun) [inline dace/sdfg/propagation.py:699]:
```suggestion
            # destination to its source. We temporarily remove this.
```

File: dace/sdfg/propagation.py [hunk 2]
```diff
@@ -789,13 +927,35 @@ def propagate_states(sdfg) -> None:
                         # Traverse as a conditional split.
                         proposed_executions = state.executions
                         proposed_dynamic = True
-                        if (getattr(state, 'full_merge_state', None) is not None
-                                and not state.dynamic_executions):
-                            full_merge_states.add(state.full_merge_state)
+
+                        # Get the dominance frontier for each child state and
+                        # merge them into one common frontier, representing the
+                        # branch's dominance frontier. If a state has no
```
Comment (tbennun) [inline dace/sdfg/propagation.py:933]:
```suggestion
                        # branch's immediate post-dominator. If a state has no
```

File: dace/sdfg/propagation.py [hunk 3]
```diff
@@ -789,13 +927,35 @@ def propagate_states(sdfg) -> None:
...
+                            frontier = dom_frontier[oedge.dst]
+                            if not frontier:
+                                frontier = {oedge.dst}
+                            common_frontier |= frontier
+
+                            # Continue traversal for each child.
                             traversal_q.append((oedge.dst, proposed_executions,
                                                 proposed_dynamic, itvar_stack))
 
+                        # If the whole branch isn't dynamic anyway, and the
```
Comment (tbennun) [inline dace/sdfg/propagation.py:948]:
```suggestion
                        # If the whole branch is not dynamic,, and the
```

File: dace/sdfg/propagation.py (hunk 2 above)
Comment (tbennun) [reply dace/sdfg/propagation.py:933]: Is it?

**Review by tbennun — CHANGES_REQUESTED** [review]

File: dace/sdfg/propagation.py (hunk 3 above)
Comment (phschaad) [reply dace/sdfg/propagation.py]: fixed in 32ff4a7

File: dace/sdfg/propagation.py (hunk 2 above)
Comment (phschaad) [reply dace/sdfg/propagation.py]: no, fixed in 32ff4a7

File: dace/sdfg/propagation.py (hunk 1 above)
Comment (phschaad) [reply dace/sdfg/propagation.py]: fixed in 32ff4a7

File: dace/sdfg/propagation.py [hunk 4]
```diff
@@ -553,16 +554,171 @@ def propagate(self, array, expressions, node_range):
...
+                # This inequality needs to be checked exactly like this due to
+                # constraints in sympy/symbolic expressions, do not simplify!!!
+                if (stride < 0) == True:
+                    rng = (stop, start, -stride)
+
+                loop_states = list(
+                    sdutils.dfs_conditional(
+                        sdfg,
+                        sources=[begin],
+                        condition=lambda _, child: child != guard))
```
[the remark this answers is not in the document]
Comment (phschaad) [reply dace/sdfg/propagation.py]: adopted in 32ff4a7

File: dace/sdfg/propagation.py [hunk 5]
```diff
@@ -553,16 +554,171 @@ def propagate(self, array, expressions, node_range):
...
+                loop_subgraph = gr.SubgraphView(sdfg, loop_states)
+                for v in loop_subgraph.nodes():
+                    v.ranges[itervar] = subsets.Range([rng])
```
[the remark this answers is not in the document]
Comment (phschaad) [reply dace/sdfg/propagation.py]: adopted in 32ff4a7

File: dace/sdfg/propagation.py [hunk 6]
```diff
@@ -553,16 +554,171 @@ def propagate(self, array, expressions, node_range):
...
+            guard_inedges = sdfg.in_edges(guard)
+            itervar = list(guard_inedges[0].data.assignments.keys())[0]
+            if itervar in begin.ranges:
+                continue
+
+            res = find_for_loop(sdfg, guard, begin)
+            if res is None:
+                unannotated_cycle_states.extend(cycle)
+            else:
+                # No range detected, mark as unbounded.
```
[the remark this answers is not in the document]
Comment (phschaad) [reply dace/sdfg/propagation.py]: fixed in 32ff4a7

**Review by tbennun — APPROVED** [review]

**Pull request merged**